packages/shared/src/hooks/post/useLeanPostActions.ts

```diff
@@ -0,0 +1,464 @@
+import { useCallback, useContext } from 'react';
+import { QueryClient, useMutation, useQueryClient } from 'react-query';
+import {
+  banPost,
+  deletePost,
+  demotePost,
+  Post,
+  promotePost,
+  updatePinnedPost,
+  UserPostVote,
+  VOTE_POST_MUTATION,
+} from '../../graphql/posts';
+import {
+  ToggleVoteProps,
+  upvoteMutationKey,
+  voteMutationHandlers,
+} from '../vote';
+import { postAnalyticsEvent } from '../../lib/feed';
+import { Origin } from '../../lib/analytics';
+import useReportPost from '../useReportPost';
+import { useAnalyticsContext } from '../../contexts/AnalyticsContext';
+import { AuthTriggers } from '../../lib/auth';
+import { graphqlUrl } from '../../lib/config';
+import { useRequestProtocol } from '../useRequestProtocol';
+import AuthContext from '../../contexts/AuthContext';
+import useBookmarkPost from '../useBookmarkPost';
+import { ToastSubject, useToastNotification } from '../useToastNotification';
+import { PromptOptions, usePrompt } from '../usePrompt';
+import { Roles } from '../../lib/user';
+import { SourcePermissions, SourceType } from '../../graphql/sources';
+import { SourceActionArguments, TagActionArguments } from '../useTagAndSource';
+import useMutateFilters from '../useMutateFilters';
+import AlertContext from '../../contexts/AlertContext';
+import useFeedSettings from '../useFeedSettings';
+
+const multiPostTransformer = (
+  posts: Post[],
+  id: string,
+  update: (oldPost: Post) => Partial<Post>,
+): Post[] =>
+  posts.map((post) =>
+    post.id === id
+      ? {
+          ...post,
+          ...update(post),
+        }
+      : post,
+  );
+
+const updateMap = {
+  'single-post': {
+    transformer: multiPostTransformer,
+  },
+  'further-reading': {
+    transformer: ({ previousData, id, update }) => ({
+      ...previousData,
+      trendingPosts: multiPostTransformer(
+        previousData.trendingPosts,
+        id,
+        update,
+      ),
+      similarPosts: multiPostTransformer(previousData.similarPosts, id, update),
+    }),
+  },
+};
+
+const updatePost =
+  (
+    queryClient: QueryClient,
+    queryKey: string[],
+    transformKey: string,
+    update: (oldPost: Post) => Partial<Post>,
+  ): ((args: { id: string }) => Promise<() => void>) =>
+  async ({ id }) => {
+    await queryClient.cancelQueries(queryKey);
+    const previousData = queryClient.getQueryData(queryKey);
+    queryClient.setQueryData(
+      queryKey,
+      updateMap[transformKey].transformer({ previousData, id, update }),
+    );
+    return () => {
+      queryClient.setQueryData(queryKey, previousData);
+    };
+  };
+
+export default function useLeanPostActions({
+  queryKey,
+  transformKey = 'further-reading',
+}) {
```

Comment:
I like the idea but I think component would need to be:
- split up into smaller parts/hooks
- based on your comment [here](https://dailydotdev.slack.com/archives/C02E2C3C13R/p1697703893665599?thread_ts=1697634724.796189&cid=C02E2C3C13R) it should support upvoting from single post as well
 - we should also see if it can utilize mutation subscriptions to sync state between different caches (as current hooks do)
 - should it solve props drilling in cards? feels like it should
 - don't get me wrong it looks great just with current functionality I think it's lacking any benefits vs current set of hooks 🙏 🚀

Reply:
Yeah 100% inline with the idea, Indeed the mutation subscription could potentially just check all caches for that postId and do something with it :)

And fully agree it should be split up again in singular logic, this was easier from a start point only.